### DISCUSSION TITLE
Response rules - Is it possible to use multiple data sources at the same time?

### DISCUSSION THREAD

**Discussion by Kieranties** [mockoon/mockoon#1440] [Help] [answered]
I have a simple API to mock that requires a valid account id to be supplied in the request url.  Additionally it requires a channel id in the querystring.  Each account can have multiple channels.

I want to mock the system so that a known number of accounts are available, and each account has pre-existing channels targeted.

To do so, I have set up a data bucket where the known account maps to a property and the property value is a list of known channels:
```json
{
  "1234": [ "5678" ]
}
```

I then have a global route configured:
```json
{
      "uuid": "4650c026-a80d-41ae-aa8a-7b54d3540555",
      "type": "http",
      "documentation": "Validate account and channel",
      "method": "all",
      "endpoint": ":account/*",
      "responses": [
        {
          "uuid": "982c9e19-bf78-474d-8d56-fcf7ae166da3",
          "body": "",
          "latency": 0,
          "statusCode": 404,
          "label": "Invalid account or channel",
          "headers": [],
          "bodyType": "INLINE",
          "filePath": "",
          "databucketID": "",
          "sendFileAsBody": false,
          "rules": [],
          "rulesOperator": "OR",
          "disableTemplating": false,
          "fallbackTo404": false,
          "default": true,
          "crudKey": "id",
          "callbacks": []
        }
      ],
      "responseMode": "FALLBACK"
    }
```

I am trying to configure a rule that will lookup the bucket path based on the account id, and the value in the list based on the channel.

I feel this would evaluate to something like
```
{{ len (filter (data 'k845' (urlParamRaw 'account')) (queryParamRaw 'channel')) 1 }}
```

I've tried various incantations but am struggling to express this in a rule.  Is it possible?

**Comment by 255kb** [comment, marked as answer]
You may not be able to use a rule as templating (access to `urlParam` helper) is not possible in the rule's target field: 
![image](https://github.com/mockoon/mockoon/assets/7489814/d97cb3a3-2d08-4709-8133-ec69e9cecc89)

This should work, combined with an `if`:

```handlebars
{{#if (includes (data 'testdata' (urlParam 'account')) (queryParamRaw 'channel'))}}
{{status 200}}
{{else}}
{{status 404}}
{{/if}}
```

PS: the `includes` helper is supposed to work only with strings, but it's working with array of primitives too, due to how JS works.

I've opened an issue to add support for templating in the rule's target: https://github.com/mockoon/mockoon/issues/1443

**Reply by Kieranties** [reply]
Ah, so your suggestion (for now) is not to use the rules, but rather process in the response body?

I'll have another play 😃

**Reply by 255kb** [reply]
yes, without being able to target dynamically the `1234` property in your bucket, you won't be able to properly test if the array contains your `channel` id.

**Comment by Kieranties** [comment]
I've just noted the related issue has been solved and released, thanks for the quick progress!
